service/src/main/java/org/ehrbase/plugin/PluginAspect.java

```diff
@@ -0,0 +1,180 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.ehrbase.plugin;
+
+import com.nedap.archie.rm.composition.Composition;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.ehrbase.api.exception.InternalServerException;
+import org.ehrbase.plugin.dto.CompositionWithEhrId;
+import org.ehrbase.plugin.extensionpoints.CompositionExtensionPointInterface;
+import org.springframework.beans.factory.ListableBeanFactory;
+import org.springframework.core.annotation.AnnotationAwareOrderComparator;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * @author Stefan Spiska
+ */
+@Component
+@Aspect
+public class PluginAspect {
+
+  public static final Comparator<Map.Entry<String, CompositionExtensionPointInterface>>
+      EXTENSION_POINTS_COMPARATOR =
+          // respect @Order
+          ((Comparator<Map.Entry<String, CompositionExtensionPointInterface>>)
+                  (e1, e2) ->
+                      AnnotationAwareOrderComparator.INSTANCE.compare(e1.getValue(), e2.getValue()))
+              .reversed()
+              // ensure constant ordering
+              .thenComparing(Map.Entry::getKey);
+
+  private static class Chain<T> {
+
+    T current;
+    Chain<T> next;
+  }
+
+  private final ListableBeanFactory beanFactory;
+
+  public PluginAspect(ListableBeanFactory beanFactory) {
+
+    this.beanFactory = beanFactory;
+  }
+
+  /**
+   * Handle Extension-points for Composition create
+   *
+   * @param pjp
+   * @return
+   * @see <a href="I_EHR_COMPOSITION in openEHR Platform Service
+   *     Model">https://specifications.openehr.org/releases/SM/latest/openehr_platform.html#_i_ehr_composition_interface</a>
+   */
+  @Around("execution(* org.ehrbase.api.service.CompositionService.create(..))")
+  public Object aroundCreateComposition(ProceedingJoinPoint pjp) {
+
+    Chain<CompositionExtensionPointInterface> chain =
+        buildChain(
+            getCompositionExtensionPointInterfaceList(),
+            new CompositionExtensionPointInterface() {});
+    CompositionWithEhrId input =
+        new CompositionWithEhrId((Composition) pjp.getArgs()[1], (UUID) pjp.getArgs()[0]);
+
+    return Optional.of(
+        handleChain(
+            chain,
+            l -> (l::aroundCreation),
+            input,
+            i -> {
+              pjp.getArgs()[1] = i.getComposition();
+              pjp.getArgs()[0] = i.getEhrId();
+
+              return ((Optional<UUID>) proceed(pjp)).orElseThrow();
```

Comment:
I think we should switch this to be something like `proceed(pjp, new Object[]{i.getEhrId(),i.getComposition()}))`

This way we would not rely on the Implementation of `ProceedingJoinPoint::getArgs` to always hand out the same instance of he array. (It currently does, but we can not assume it always will)